android/src/main/java/io/radar/react/RNRadarModule.java

```
@@ -109,29 +163,69 @@ public void updateLocation(ReadableMap locationMap, final Promise promise) {
+                try {
+                    if (status == Radar.RadarStatus.SUCCESS) {
+                        WritableMap map = Arguments.createMap();
+                        map.putString("status", status.toString());
+                        if (location != null) {
+                            map.putMap("location", RNRadarUtils.mapForJson(Radar.jsonForLocation(location)));
+                        }
+                        if (events != null) {
+                            map.putArray("events", RNRadarUtils.arrayForJson(RadarEvent.toJson(events)));
+                        }
+                        if (user != null) {
+                            map.putMap("user", RNRadarUtils.mapForJson(user.toJson()));
+                        }
+                        promise.resolve(map);
+                    } else {
+                        promise.reject(status.toString(), status.toString());
+                    }
+                } catch (JSONException e) {
+                    promise.reject(Radar.RadarStatus.ERROR_SERVER.toString(), Radar.RadarStatus.ERROR_SERVER.toString());
                 }
             }
         });
     }
 
+    @ReactMethod
+    public void startTrackingEfficient() {
+        Radar.startTracking(RadarTrackingOptions.EFFICIENT);
+    }
+
+    @ReactMethod
+    public void startTrackingResponsive() {
+        Radar.startTracking(RadarTrackingOptions.RESPONSIVE);
+    }
+
+    @ReactMethod
+    public void startTrackingContinuous() {
+        Radar.startTracking(RadarTrackingOptions.CONTINUOUS);
+    }
+
+    @ReactMethod
+    public void startTrackingCustom(ReadableMap optionsMap) {
+        try {
+            JSONObject optionsObj = RNRadarUtils.jsonForMap(optionsMap);
+            RadarTrackingOptions options = RadarTrackingOptions.fromJson(optionsObj);
+            Radar.startTracking(options);
+        } catch (JSONException e) {
```

is it possible for `json` serialization of an `optionsMap` to error out?

---

Yeah, the `optionsMap` methods can throw `JSONException`. (I'm not sure why React Native uses these instead of `JSONObject`s in the first place...)

---

yeah, definitely odd, let's bubble up this exception to the caller then?